Update some code before IAEA workshop

GiudGiud opened the pull request:
This adds enough capability to do the CHT example with FV by:
- avoiding a seg fault on using diffusionfluxaux with FV

also closes #18061

More importantly:
- it fixes the heat source restriction in the heat conduction physics
- it lets you tinker with the variables to relax in the fixed point solve when using Picard
- it errors when your fixed-point-iteration setup is not going to relax or transform anything

GiudGiud committed:
Add errors when relaxation is not doing anything

GiudGiud committed:
Begin to add support for relaxing the auxiliary system in fixed point iterations
- pick the right system
Note that this is not enough because auxvars could be updated on timestep begin/end after the transformations
It s very tricky to have something that always works

refs #20100

GiudGiud committed:
Fix block restriction in heat conduction physics
refs #25642

GiudGiud committed:
Add support for finite volume for the diffusion flux aux
refs #31566

GiudGiud committed:
Move finding the system into allocateStorage because storage is allocated before initial setup for secondary transforms
refs #20100

GiudGiud committed:
Update tests that removed the transformed PPs

GiudGiud committed:
Add more criteria for triggering fixed point iterations

GiudGiud committed:
Add relaxation to tests that used transformation of the secondary PPs: no regold, jsut more iterations

GiudGiud committed:
Keep track of which systems to keep the previous value of between solver & aux systems

GiudGiud commented on test/tests/multiapps/picard_postprocessor/tests:
```diff
@@ -94,7 +94,11 @@
-      cli_args = "Outputs/file_base=app_begin_transfer_begin_picard_sub Executioner/transformed_postprocessors='' MultiApps/sub/transformed_postprocessors='from_main'"
+      cli_args = "Outputs/file_base=app_begin_transfer_begin_picard_sub
+                  Executioner/transformed_postprocessors=''
+                  Executioner/relaxation_factor=1
```
this is needed to avoid a new error with the line above

GiudGiud committed:
Remove relaxation parameter where it was not used

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3238301/), step Docs: sync website on 631bf16 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31572/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3238295/), step Generate coverage on 631bf16 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>fd74ac</th>
      <th colspan=4>#31572 631bf1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.90%</td>
      <td>85.90%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>122277</td>
      <td>122331</td>
      <td>+54</td>
      <td>108</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>20078</td>
      <td>20086</td>
      <td>+8</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31572/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31572/coverage/framework)


## Modules coverage

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>fd74ac</th>
      <th colspan=4>#31572 631bf1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.19%</td>
      <td>89.12%</td>
      <td>-0.06%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4611</td>
      <td>4613</td>
      <td>+2</td>
      <td>10</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>559</td>
      <td>563</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31572/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31572/coverage/heat_transfer)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31572/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented:
can you re-push so we can have up-to-date testing?

lindsayad commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -2353,6 +2353,30 @@ class FEProblemBase : public SubProblem, public Restartable
+   * Set a flag that indicated that user required values for the previous multiapp fixed point
+   * iterate
+   */
+  void needsPreviousMultiAppFixedPointIterationSolver(bool state);
```
```suggestion
  void needsPreviousMultiAppFixedPointIterationSolution(bool state);
```
?

GiudGiud replied:
it could use a solver system number but then I could see a need for relaxing two variables in two different system and then it would be a vector of solver system numbers instead

lindsayad commented on framework/src/executioners/FixedPointSolve.C:
```diff
@@ -589,3 +614,36 @@ FixedPointSolve::performingRelaxation(const bool primary) const
+    if (_problem.hasAuxiliaryVariable(transformed_vars[0]))
+      _transformed_sys = &_aux;
+    else
+      _transformed_sys = &_solver_sys;
```
So we copy previous fixed point solutions for all systems but then only do transformations for one?

GiudGiud committed:
Address review:
- rename API Solver->Solution for the solver systems
- dont trigger fixed point iterations silently, instead error on missing parameters
- fix some docstring

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud committed:
Address Alex's review:
- add system number to APIs that keep track of the need for an old multiapp fixed point state

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3582912/), step Results summary on 72be3ab wanted to post the following:

## Framework test summary

Compared against fb4ac69 in job [civet.inl.gov/job/3577491](https://civet.inl.gov/job/3577491).

### No added tests


### Run time changes

| Test                                                                               | Base (s)   | Head (s)   | +/-     | Base (MB)   | Head (MB)   |
|------------------------------------------------------------------------------------|------------|------------|---------|-------------|-------------|
| `multiapps/picard_postprocessor.pp_transient/app_end_transfers_end_picard_sub`     | 2.88       | 5.02       | +74.62% | 0.00        | 0.00        |
| `multiapps/picard_postprocessor.pp_transient/app_begin_transfers_begin_picard_sub` | 3.12       | 4.85       | +55.35% | 0.00        | 0.00        |
## Modules test summary

Compared against fb4ac69 in job [civet.inl.gov/job/3577491](https://civet.inl.gov/job/3577491).

No change

GiudGiud commented:
Test failures unrelated
